dali/kernels/transpose/transpose_gpu.cu

```diff
@@ -0,0 +1,365 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/kernels/transpose/transpose_gpu.h"
+#include <cuda_runtime.h>
+#include <memory>
+#include <vector>
+#include "dali/core/util.h"
+#include "dali/kernels/transpose/transpose_gpu_def.h"
+#include "dali/kernels/transpose/transpose_gpu_impl.cuh"
+#include "dali/kernels/transpose/transpose_gpu_setup.cuh"
+
+namespace dali {
+namespace kernels {
+namespace transpose_impl {
+
+struct TransposeInfo {
+  int                 element_size;
+  TransposeMethod     method;
+  TensorShape<>       shape;
+  SmallVector<int, 6> perm;
+};
+
+constexpr int kMaxInterleaveSize = 32;
+constexpr int kMaxDeinterleaveSize = kMaxInterleaveSize;
+
+inline bool UseTiledTranspose(const int64_t *shape, const int *perm, int ndim) {
+  int xdim = ndim-1;
+  int ydim = 0;
+  for (; ydim < xdim; ydim++) {
+    if (perm[ydim] == xdim)
+      break;
+  }
+  double tile_coverage = shape[xdim] * shape[ydim];
+  tile_coverage /= align_up(shape[xdim], kTileSize) * align_up(shape[ydim], kTileSize);
+  return tile_coverage > 0.4;  // for now, it's an educated guess
+}
+
+TransposeMethod GetTransposeMethod(const int64_t *shape,
+                                   const int *perm,
+                                   int ndim,
+                                   int element_size) {
+  if (ndim == 1)
+    return TransposeMethod::Copy;
+  if (perm[ndim-1] == ndim - 1) {
+    assert(ndim >= 3);
+    if (shape[ndim-1] * element_size < kTiledTransposeMaxVectorSize) {
+      if (UseTiledTranspose(shape, perm, ndim-1))
```

Comment:
wondering: why not encapsulate the whole "should I use tiled transpose" inside `UseTiledTranspose`?